### PR TITLE
Unwrap dynamic error types when inner is simple `PyErr`

### PR DESCRIPTION
**Pull request by BlueGlassBlock**
This is the first part of suggested improvements in #2998.

This change will make bubbled `PyErr` wrapped in `eyre::Report` / `anyhow::Error` bubble up unchanged, instead of being wrapped in a `PyRuntimeError`.

### CODE REVIEW

**Comment by adamreichold** [comment]
@mejrs Will wait for your feedback before merging as you sounded unsure whether this is a sensible thing to do in the issue.

File: src/conversions/anyhow.rs
Comment (davidhewitt) [inline src/conversions/anyhow.rs:117]: Rather than the assignment here, might it be simpler to just do `if let Ok(py_err) = error.downcast::<Self>() { return py_err; }`

And same for the eyre implementation.

File: src/conversions/anyhow.rs
Comment (davidhewitt) [inline src/conversions/anyhow.rs:115]: This comment is slightly confusing to me as a reader; why are they skipped?

I think the following would be clearer:

```suggestion
        // Errors containing a PyErr without chain or context are returned as the underlying error 
```

**Review by davidhewitt — COMMENTED** [review]
While I think this change is strictly an improvement, I'd like to quickly take a moment to think about how we document this and / or provide backwards compatibility.

Changing the error type observed here may catch some users out. Probably module docs need updating. Should this go in the migration guide?

File: src/conversions/anyhow.rs
Comment (adamreichold) [reply src/conversions/anyhow.rs:117]: This does not work as `downcast` consumes `error`, but if the downcast fails, we still want to format it into a string. So either we duplicate that code path or use the assignment. Personally, I find the assignment to be an elegant solution to this problem.

Comment (davidhewitt) [reply src/conversions/anyhow.rs:117]: Ah great point, of course. Ignore me!

**Comment by DataTriny** [comment]
Hello @BlueGlassBlock 

Thank you for contributing to PyO3!

The project is undergoing changes on its licensing model to better align with the rest of the Rust crates ecosystem.

Before your work gets merged, please express your consent by leaving a comment on this [pull request](https://github.com/PyO3/pyo3/pull/3108).

Thank you!

**Comment by mejrs** [comment]
> @mejrs Will wait for your feedback before merging as you sounded unsure whether this is a sensible thing to do in the issue.

I have nothing, feel free to merge as you wish :)

**Comment by adamreichold** [comment]
@BlueGlassBlock You would be up to writing an entry for the migration guide showing the difference when forwarding errors using `anyhow` or `eyre`? The module docs definitely must be changed to match the proposed implementation.

**Comment by BlueGlassBlock** [comment]
> @BlueGlassBlock You would be up to writing an entry for the migration guide showing the difference when forwarding errors using `anyhow` or `eyre`? The module docs definitely must be changed to match the proposed implementation.

Certainly!

File: guide/src/migration.md
Comment (adamreichold) [inline guide/src/migration.md:27]: I think it would be nice to just define a `#[pyfunction]` in the example and call that using `wrap_pyfunction!` and `py_run!` instead of hinting at this in a comment.

File: guide/src/migration.md
Comment (adamreichold) [inline guide/src/migration.md:20]: While this works, it is bit confusing, maybe better do

```rust
Err(PyValueError::new_err("original error message").into())
```

**Comment by adamreichold** [comment]
Thanks for drafting the migration guide entry. Please remember to adjust the existing module-level docs in `anyhow.rs` and `eyre.rs` which currently still state that a `RuntimeError` is produced unconditionally.

File: guide/src/migration.md
Comment (adamreichold) [inline guide/src/migration.md:39]:
```suggestion
After, the same code will print `ValueError: original error message`, which is more straightforward.
```

**Comment by adamreichold** [comment]
> Changing the error type observed here may catch some users out. Probably module docs need updating. Should this go in the migration guide?

@davidhewitt I think this better positioned now w.r.t. the above concerns and therefore ready to be merged. Do you agree?

**Review by davidhewitt — APPROVED** [review]
Yep this looks great to me now, thanks! I'll squash and merge this now.

**Comment by davidhewitt** [comment]
bors r=davidhewitt,adamreichold,mejrs

**Pull request merged**
